File: src/scheduler.ts
import { execFileSync, execSync } from 'node:child_process'
import { existsSync, unlinkSync, writeFileSync } from 'node:fs'

import { log } from './logger.js'
import { getBinaryInstallPath, getPlatform } from './platform.js'

const MACOS_PLIST_PATH = '/Library/LaunchDaemons/com.glean.mdm.plist'
const LINUX_SERVICE_PATH = '/etc/systemd/system/glean-mdm.service'
const LINUX_TIMER_PATH = '/etc/systemd/system/glean-mdm.timer'
const WINDOWS_TASK_NAME = 'Glean MDM'

/** Random minute (0–59) to stagger scheduled runs and avoid thundering-herd on the version endpoint. */
export function randomMinute(): number {
  return Math.floor(Math.random() * 60)
}

/** Exposed for tests — argv array avoids shell quoting bugs when paths contain spaces. */
export function schtasksCreateArgs(binaryPath: string, minute: number): string[] {
  const startTime = `09:${String(minute).padStart(2, '0')}`
  return ['/Create', '/TN', WINDOWS_TASK_NAME, '/TR', `${binaryPath} run`, '/SC', 'DAILY', '/ST', startTime, '/RU', 'SYSTEM', '/F']
}

export function buildMacOSPlist(binaryPath: string, minute: number): string {
  return `<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>com.glean.mdm</string>
    <key>ProgramArguments</key>
    <array>
        <string>${binaryPath}</string>
        <string>run</string>
    </array>
    <key>StartCalendarInterval</key>
    <dict>
        <key>Hour</key>
        <integer>9</integer>
        <key>Minute</key>
        <integer>${minute}</integer>
    </dict>
    <key>RunAtLoad</key>
    <true/>
</dict>
</plist>`
}

function installMacOSSchedule(): void {
  const binaryPath = getBinaryInstallPath()
  const minute = randomMinute()
  const plist = buildMacOSPlist(binaryPath, minute)

  writeFileSync(MACOS_PLIST_PATH, plist)
  try {
    execSync(`launchctl bootout system "${MACOS_PLIST_PATH}"`, { stdio: 'ignore' })
  } catch {
    // May not be loaded
  }
  execSync(`launchctl bootstrap system "${MACOS_PLIST_PATH}"`)
  log.info(`Installed macOS LaunchDaemon schedule (daily at 9:${String(minute).padStart(2, '0')} AM)`)
}

function uninstallMacOSSchedule(): void {
  const existed = existsSync(MACOS_PLIST_PATH)
  try {
    execSync(`launchctl bootout system "${MACOS_PLIST_PATH}"`, { stdio: 'ignore' })
  } catch {
    // May not be loaded
  }
  try {
    unlinkSync(MACOS_PLIST_PATH)
  } catch {
    // May not exist
  }
  if (existed) {
    log.info('Removed macOS LaunchDaemon schedule')
  }
}

function installLinuxSchedule(): void {
  const binaryPath = getBinaryInstallPath()

  const service = `[Unit]
Description=Glean MDM
After=network-online.target
Wants=network-online.target

[Service]
Type=oneshot
ExecStart=${binaryPath} run

[Install]
WantedBy=multi-user.target
`

  const minute = randomMinute()
  const timer = `[Unit]
Description=Daily Glean MDM

[Timer]
OnCalendar=*-*-* 09:${String(minute).padStart(2, '0')}:00
Persistent=true

[Install]
WantedBy=timers.target
`

  writeFileSync(LINUX_SERVICE_PATH, service)
  writeFileSync(LINUX_TIMER_PATH, timer)
  execSync('systemctl daemon-reload')
  execSync('systemctl enable --now glean-mdm.timer')
  log.info(`Installed systemd timer schedule (daily at 9:${String(minute).padStart(2, '0')} AM)`)
}

function uninstallLinuxSchedule(): void {
  const existed = existsSync(LINUX_SERVICE_PATH) || existsSync(LINUX_TIMER_PATH)
  try {
    execSync('systemctl disable --now glean-mdm.timer', {
      stdio: 'ignore',
    })
  } catch {
    // May not be enabled
  }
  try {
    unlinkSync(LINUX_SERVICE_PATH)
  } catch {
    // May not exist
  }
  try {
    unlinkSync(LINUX_TIMER_PATH)
  } catch {
    // May not exist
  }
  try {
    execSync('systemctl daemon-reload', { stdio: 'ignore' })
  } catch {
    // Best effort
  }
  if (existed) {
    log.info('Removed systemd timer schedule')
  }
}

function installWindowsSchedule(): void {
  const binaryPath = getBinaryInstallPath()
  const minute = randomMinute()
  execFileSync('schtasks', schtasksCreateArgs(binaryPath, minute))
  // Enable catch-up: run the task if a scheduled run was missed while the machine was off
  execSync(
    `powershell -Command "$t = Get-ScheduledTask '${WINDOWS_TASK_NAME}'; $t.Settings.StartWhenAvailable = $true; $t | Set-ScheduledTask"`,
  )
  log.info(`Installed Windows Task Scheduler schedule (daily at 9:${String(minute).padStart(2, '0')} AM)`)
}

function uninstallWindowsSchedule(): void {
  let removed = false
  try {
    execSync(`schtasks /Delete /TN "${WINDOWS_TASK_NAME}" /F`, {
      stdio: 'ignore',
    })
    removed = true
  } catch {
    // May not exist
  }
  if (removed) {
    log.info('Removed Windows Task Scheduler schedule')
  }
}

export function installSchedule(): void {
  switch (getPlatform()) {
    case 'darwin':
      installMacOSSchedule()
      break
    case 'linux':
      installLinuxSchedule()
      break
    case 'win32':
      installWindowsSchedule()
      break
  }
}

export function uninstallSchedule(): void {
  switch (getPlatform()) {
    case 'darwin':
      uninstallMacOSSchedule()
      break
    case 'linux':
      uninstallLinuxSchedule()
      break
    case 'win32':
      uninstallWindowsSchedule()
      break
  }
}

File: internal/scheduler/scheduler.go
// Package scheduler installs/removes the system scheduled task (launchd,
// systemd, or Task Scheduler), mirroring scheduler.ts.
package scheduler

import (
	cryptorand "crypto/rand"
	"fmt"
	"math/big"
	"os"
	"os/exec"

	"github.com/gleanwork/glean-mdm/internal/logger"
	"github.com/gleanwork/glean-mdm/internal/platform"
)

const (
	macOSPlistPath   = "/Library/LaunchDaemons/com.glean.mdm.plist"
	linuxServicePath = "/etc/systemd/system/glean-mdm.service"
	linuxTimerPath   = "/etc/systemd/system/glean-mdm.timer"
	windowsTaskName  = "Glean MDM"
)

// RandomMinute returns a random minute (0-59) to stagger scheduled runs.
func RandomMinute() int {
	n, err := cryptorand.Int(cryptorand.Reader, big.NewInt(60))
	if err != nil {
		return 0
	}
	return int(n.Int64())
}

// SchtasksCreateArgs builds the schtasks /Create arguments. The /TR value must
// quote paths with spaces for Task Scheduler.
func SchtasksCreateArgs(binaryPath string, minute int) []string {
	startTime := fmt.Sprintf("09:%02d", minute)
	return []string{"/Create", "/TN", windowsTaskName, "/TR", fmt.Sprintf("\"%s\" run", binaryPath), "/SC", "DAILY", "/ST", startTime, "/RU", "SYSTEM", "/F"}
}

// BuildMacOSPlist renders the LaunchDaemon plist.
func BuildMacOSPlist(binaryPath string, minute int) string {
	return fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>com.glean.mdm</string>
    <key>ProgramArguments</key>
    <array>
        <string>%s</string>
        <string>run</string>
    </array>
    <key>StartCalendarInterval</key>
    <dict>
        <key>Hour</key>
        <integer>9</integer>
        <key>Minute</key>
        <integer>%d</integer>
    </dict>
    <key>RunAtLoad</key>
    <true/>
</dict>
</plist>`, binaryPath, minute)
}

func installMacOSSchedule() {
	binaryPath := platform.GetBinaryInstallPath()
	minute := RandomMinute()
	plist := BuildMacOSPlist(binaryPath, minute)

	if err := os.WriteFile(macOSPlistPath, []byte(plist), 0o644); err != nil {
		logger.Error("Failed to write LaunchDaemon plist: %v", err)
		return
	}
	_ = exec.Command("launchctl", "bootout", "system", macOSPlistPath).Run()
	if err := exec.Command("launchctl", "bootstrap", "system", macOSPlistPath).Run(); err != nil {
		logger.Error("Failed to bootstrap LaunchDaemon: %v", err)
		return
	}
	logger.Info("Installed macOS LaunchDaemon schedule (daily at 9:%02d AM)", minute)
}

func uninstallMacOSSchedule() {
	_, statErr := os.Stat(macOSPlistPath)
	existed := statErr == nil
	_ = exec.Command("launchctl", "bootout", "system", macOSPlistPath).Run()
	_ = os.Remove(macOSPlistPath)
	if existed {
		logger.Info("Removed macOS LaunchDaemon schedule")
	}
}

func installLinuxSchedule() {
	binaryPath := platform.GetBinaryInstallPath()

	service := fmt.Sprintf(`[Unit]
Description=Glean MDM
After=network-online.target
Wants=network-online.target

[Service]
Type=oneshot
ExecStart=%s run

[Install]
WantedBy=multi-user.target
`, binaryPath)

	minute := RandomMinute()
	timer := fmt.Sprintf(`[Unit]
Description=Daily Glean MDM

[Timer]
OnCalendar=*-*-* 09:%02d:00
Persistent=true

[Install]
WantedBy=timers.target
`, minute)

	if err := os.WriteFile(linuxServicePath, []byte(service), 0o644); err != nil {
		logger.Error("Failed to write systemd service: %v", err)
		return
	}
	if err := os.WriteFile(linuxTimerPath, []byte(timer), 0o644); err != nil {
		logger.Error("Failed to write systemd timer: %v", err)
		return
	}
	_ = exec.Command("systemctl", "daemon-reload").Run()
	if err := exec.Command("systemctl", "enable", "--now", "glean-mdm.timer").Run(); err != nil {
		logger.Error("Failed to enable systemd timer: %v", err)
		return
	}
	logger.Info("Installed systemd timer schedule (daily at 9:%02d AM)", minute)
}

func uninstallLinuxSchedule() {
	_, serviceErr := os.Stat(linuxServicePath)
	_, timerErr := os.Stat(linuxTimerPath)
	existed := serviceErr == nil || timerErr == nil

	_ = exec.Command("systemctl", "disable", "--now", "glean-mdm.timer").Run()
	_ = os.Remove(linuxServicePath)
	_ = os.Remove(linuxTimerPath)
	_ = exec.Command("systemctl", "daemon-reload").Run()
	if existed {
		logger.Info("Removed systemd timer schedule")
	}
}

func installWindowsSchedule() {
	binaryPath := platform.GetBinaryInstallPath()
	minute := RandomMinute()
	if err := exec.Command("schtasks", SchtasksCreateArgs(binaryPath, minute)...).Run(); err != nil {
		logger.Error("Failed to create Windows scheduled task: %v", err)
		return
	}
	// Enable catch-up: run the task if a scheduled run was missed while the machine was off.
	script := fmt.Sprintf(`$t = Get-ScheduledTask '%s'; $t.Settings.StartWhenAvailable = $true; $t | Set-ScheduledTask`, windowsTaskName)
	_ = exec.Command("powershell", "-Command", script).Run()
	logger.Info("Installed Windows Task Scheduler schedule (daily at 9:%02d AM)", minute)
}

func uninstallWindowsSchedule() {
	if err := exec.Command("schtasks", "/Delete", "/TN", windowsTaskName, "/F").Run(); err != nil {
		logger.Error("Failed to uninstall Windows schedule: %v", err)
		return
	}
	logger.Info("Removed Windows Task Scheduler schedule")
}

// Install installs the scheduled task for the current platform.
func Install() {
	switch platform.Get() {
	case platform.Darwin:
		installMacOSSchedule()
	case platform.Linux:
		installLinuxSchedule()
	case platform.Win32:
		installWindowsSchedule()
	}
}

// Uninstall removes the scheduled task for the current platform.
func Uninstall() {
	switch platform.Get() {
	case platform.Darwin:
		uninstallMacOSSchedule()
	case platform.Linux:
		uninstallLinuxSchedule()
	case platform.Win32:
		uninstallWindowsSchedule()
	}
}
